### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by reebok159**
http://egnbookstore.herokuapp.com
Admin:
admin@loc.loc:123456q

test coupon: zzafg-8

### CODE REVIEW

File: app/forms/address_form.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  def build_address
+    @type = address_type
+    return unless @type
+    @current_user.send(:"build_#{@type}", @params[@type])
+  end
+
+  def address_type
+    types = %w[billing_address shipping_address]
+    @params.keys.find { |item| types.include? item }
```
[the remark this answers is not in the document]
Comment (reebok159) [reply app/forms/address_form.rb]: what you mean?

File: app/helpers/orders_helper.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module OrdersHelper
+  def order_filters
+    Order.statuses.select { |_, val| val.positive? }
```
[the remark this answers is not in the document]
Comment (reebok159) [reply app/helpers/orders_helper.rb]: what you mean?

File: app/controllers/checkout_controller.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  end
+
+  def change_state
+    @form.order.update(checkout_state: :confirm) && return if return_to_confirm?
+    @form.order.next_state!
+    save_order_for_last_step if step == :confirm
+  end
+
+  def return_to_confirm?
+    return true if params.key?(:edit) && @form.filled?
```
Comment (dzemlianoi-double) [inline app/controllers/checkout_controller.rb]: params.key?(:edit) && @form.filled?

File: app/controllers/checkout_controller.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  end
+
+  def redirect_if_empty
+    redirect_to cart_page_path, flash: { alert: t('checkout.emptycart') } if active_order.order_items.blank?
+  end
+
+  def goto_active_step
+    @form.order.update(checkout_state: step) if params.key?(:edit)
+    jump_to(@form.order.checkout_state) unless step.to_s == @form.order.checkout_state
```
Comment (dzemlianoi-double) [inline app/controllers/checkout_controller.rb]: if step.to_s != @form.order.checkout_state
don't use unless where you can skip it

File: app/forms/checkout_form.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  def initialize(order)
+    @order = order
+  end
+
+  def update(params, step)
+    @params = params
+    case step
+    when :address then processing_address
+    when :delivery
+      @order.delivery_method = DeliveryMethod.find_by(id: @params[:delivery_method])
```
Comment (dzemlianoi-double) [inline app/forms/checkout_form.rb]: move delivery and payment case-when branches to separate methods, please

File: app/forms/checkout_form.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    @order.total_price = @order.pre_total_price
+    @order.status = :in_queue
+    @order.save_price_items
+    @order.save
+  end
+
+  def deactivate_coupon
+    return unless @order.coupon
+
+    @order.coupon.deactivate
```
Comment (dzemlianoi-double) [inline app/forms/checkout_form.rb]: @order.coupon&.deactivate ?

File: app/forms/checkout_form.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      return
+    end
+    @order.build_shipping_address(@params[:shipping_address])
+    @order.use_billing = nil
+  end
+
+  def processing_confirm
+    @order.completed_at = Time.current
+    @order.total_price = @order.pre_total_price
+    @order.status = :in_queue
```
Comment (dzemlianoi-double) [inline app/forms/checkout_form.rb]: if you use AASM - use transitions of AASM

File: app/models/book.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  has_and_belongs_to_many :authors
+
+  validates :name, :price, presence: true
+
+  scope :latest, ->(num = 2) { last(num) }
+  scope :select_category, ->(cat) { where(category: cat) if cat.present? }
+  scope :select_sort, ->(val) { order(SORT_PARAMS[val]) if SORT_PARAMS.key?(val) }
+  scope :search, ->(val) { where('lower(name) LIKE ?', "%#{val.downcase}%") if val.present? }
+
+  def self.bestsellers(num = 4)
```
Comment (dzemlianoi-double) [inline app/models/book.rb]: 4? 
Move to constant

File: app/models/order.rb [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  def total_quantity
+    order_items.pluck(:quantity).sum
+  end
+
+  def subtotal
+    order_items.collect { |item| item.quantity * item.book.price }.sum
+  end
+
+  def coupon_discount
+    coupon&.discount || 0
```
Comment (dzemlianoi-double) [inline app/models/order.rb]: coupon&.discount.to_i

File: app/models/order.rb [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  end
+
+  def delivery_price
+    delivery_method&.cost || 0
```
Comment (dzemlianoi-double) [inline app/models/order.rb]: delivery_method&.cost.to_i

File: app/models/order.rb [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+  def pre_total_price
+    subtotal + delivery_price - coupon_discount
+  end
+
+  def check_need_coupon
+    update(coupon: nil) if total_quantity.zero? && coupon
+  end
+
+  def merge_order_items(order2)
```
Comment (dzemlianoi-double) [inline app/models/order.rb]: don't use numbers in variable names in such way

File: app/models/order_item.rb [hunk 12]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class OrderItem < ApplicationRecord
+  belongs_to :order
+  belongs_to :book
+
+  validates :quantity, numericality: { greater_than_or_equal_to: 0 }
+
+  after_destroy do |item|
+    item&.order&.check_need_coupon
```
Comment (dzemlianoi-double) [inline app/models/order_item.rb]: try not to use too many try's

**Pull request merged**